### PR TITLE
Check conditional types in depth in `isTypeParameterAtTopLevel`

### LINKED ISSUE #52620: Generic not widen when union and conditional are involved.
# Bug Report

I've read #10676, but I'm failing to understand why there is no consistent widening accross the 3 examples mentioned below. 

### 🔎 Search Terms

* literal 
* widening
* union
* generics

### 🕗 Version & Regression Information

AFAIK not a regression. 

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiD2AnAtgYXgOwC6PiAPACoB80A3gLABQ00wmEOArsFkgBRYBch0ApgA8sfDABMYTMXwBmASwx9R0APzQFANz6JoXaIQCU5AL7UTVehkbQw0ALxq+AdzhI0mHHnYByL-oDc1AD0gbS0AHrK1NSgkDAIKOjYuCAATESklDR0DMysHNyEAD6MiPIA5oZkZmYWVgBGdg7O8W5JeCnsAESd-kEhoRFR5uBQLgnuyQDM6eTUtLW5bIicPPxCIuLQkqIy8ooqDprauuxFJeX6ldXROXSNCs2uiR4gk109AVTBodCDVNRAA)

### 💻 Code

```ts
class FormControl<T> {
  constructor(t:T extends undefined ? never : T) {}
}
const a = new FormControl('');
//   ^? FormControl<string>

class FormControl2<T> {
  constructor(t:T|string) {}
}
const b = new FormControl2("");
//    ^? FormControl<string>

class FormControl3<T> {
  constructor(t:T extends undefined ? never : (T|string)) {}
}
const c = new FormControl3("");
//    ^? FormControl<"">
```

### 🙁 Actual behavior

Non consistant behaviour between use-cases. 

### 🙂 Expected behavior

Consistent behaviour.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #52620.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
No changes in tests or performance. This should be ready to merge.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23667,10 +23667,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
     }
 
-    function isTypeParameterAtTopLevel(type: Type, typeParameter: TypeParameter): boolean {
-        return !!(type === typeParameter ||
-            type.flags & TypeFlags.UnionOrIntersection && some((type as UnionOrIntersectionType).types, t => isTypeParameterAtTopLevel(t, typeParameter)) ||
-            type.flags & TypeFlags.Conditional && (getTrueTypeFromConditionalType(type as ConditionalType) === typeParameter || getFalseTypeFromConditionalType(type as ConditionalType) === typeParameter));
+    function isTypeParameterAtTopLevel(type: Type, tp: TypeParameter, depth = 0): boolean {
+        return !!(type === tp ||
+            type.flags & TypeFlags.UnionOrIntersection && some((type as UnionOrIntersectionType).types, t => isTypeParameterAtTopLevel(t, tp, depth)) ||
+            depth < 3 && type.flags & TypeFlags.Conditional && (
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:23673]: What is the depth check ensuring?

Comment (ahejlsberg) [reply src/compiler/checker.ts:23673]: That we don't infinitely recurse in conditional types that reference themselves.

Comment (ahejlsberg) [reply src/compiler/checker.ts:23673]: There are examples of such types in the DT typings for `lodash`.

Comment (jakebailey) [reply src/compiler/checker.ts:23673]: The test added in the PR seems to still pass if I remove `depth < 3`; is there a test we can add which hits this so we don't change this value in a bad way later?

Comment (ahejlsberg) [reply src/compiler/checker.ts:23673]: Latest commit adds tests up to three levels of conditional types.

**Review by weswigham — APPROVED** [review]
Might be a good idea to add a 4th level of nesting to the test along with a comment saying that the weird difference in behavior that one'll have is due to this depth limit, in case someone comes by and changes it down or up in the future for some reason.

**Pull request merged**